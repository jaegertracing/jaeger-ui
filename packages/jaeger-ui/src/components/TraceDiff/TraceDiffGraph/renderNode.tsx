// Copyright (c) 2018-2020 The Jaeger Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { Popover } from 'antd';
import cx from 'classnames';
import { TLayoutVertex } from '@jaegertracing/plexus/lib/types';

import EmphasizedNode from '../../common/EmphasizedNode';
import CopyIcon from '../../common/CopyIcon';
import { TDiffCounts } from '../../../model/trace-dag/types';
import TDagPlexusVertex from '../../../model/trace-dag/types/TDagPlexusVertex';

import './renderNode.css';

type Props = {
  a: number;
  b: number;
  operation: string;
  service: string;
};

const abs = Math.abs;
const max = Math.max;

export class DiffNode extends React.PureComponent<Props> {
  render() {
    const { a, b, operation, service } = this.props;
    const isSame = a === b;
    const className = cx({
      'is-same': isSame,
      'is-changed': !isSame,
      'is-more': b > a && a > 0,
      'is-added': a === 0,
      'is-less': a > b && b > 0,
      'is-removed': b === 0,
    });
    const chgSign = a < b ? '+' : '-';
    const table = (
      <table className={`DiffNode ${className}`}>
        <tbody className="DiffNode--body">
          <tr>
            <td className={`DiffNode--metricCell ${className}`} rowSpan={isSame ? 2 : 1}>
              {isSame ? null : <span className="DiffNode--metricSymbol">{chgSign}</span>}
              {isSame ? a : abs(b - a)}
            </td>
            <td className={`DiffNode--labelCell ${className}`}>
              <strong>{service}</strong>
              <CopyIcon
                className="DiffNode--copyIcon"
                copyText={`${service} ${operation}`}
                tooltipTitle="Copy label"
              />
            </td>
          </tr>
          <tr>
            {isSame ? null : (
              <td className={`DiffNode--metricCell ${className}`}>
                <span className="DiffNode--metricSymbol">{chgSign}</span>
                {a === 0 || b === 0 ? 100 : abs(((a - b) / max(a, b)) * 100).toFixed(0)}
                <span className="DiffNode--metricSymbol">%</span>
              </td>
            )}
            <td className={`DiffNode--labelCell ${className}`}>{operation}</td>
          </tr>
        </tbody>
      </table>
    );

    return (
      <Popover overlayClassName={`DiffNode--popover ${className}`} mouseEnterDelay={0.25} content={table}>
        {table}
      </Popover>
    );
  }
}

export default function renderNode(vertex: TDagPlexusVertex<TDiffCounts>) {
  const { a, b, operation, service } = vertex.data;
  const lenA = a ? a.length : 0;
  const lenB = b ? b.length : 0;
  return <DiffNode a={lenA} b={lenB} operation={operation} service={service} />;
}

export function getNodeEmphasisRenderer(keys: Set<string>) {
  return function drawEmphasizedNode(lv: TLayoutVertex<any>) {
    if (!keys.has(lv.vertex.key)) {
      return null;
    }
    return <EmphasizedNode height={lv.height} width={lv.width} />;
  };
}
