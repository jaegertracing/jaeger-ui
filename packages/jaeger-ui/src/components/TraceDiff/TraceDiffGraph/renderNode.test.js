// Copyright (c) 2019 The Jaeger Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React from 'react';
import { render, cleanup, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import renderNode, { DiffNode, getNodeEmphasisRenderer } from './renderNode';
import EmphasizedNode from '../../common/EmphasizedNode';

describe('drawNode', () => {
  const operation = 'operationName';
  const service = 'serviceName';
  const defaultCount = 100;

  afterEach(cleanup);

  describe('diffNode', () => {
    const baseProps = {
      a: 100,
      b: 100,
      operation: 'operationName',
      service: 'serviceName',
    };

    it('renders table when a and b are the same', () => {
      render(<DiffNode {...baseProps} />);
      const metricCell = screen.getByTestId('diff-metric-cell');
      expect(metricCell).toHaveClass('is-same');
      expect(screen.getByRole('table')).toBeInTheDocument();
      expect(screen.getByText('serviceName')).toBeInTheDocument();
      expect(screen.getByText('operationName')).toBeInTheDocument();
    });

    it('applies is-more class when a < b', () => {
      render(<DiffNode {...baseProps} a={50} b={100} />);
      const metricCell = screen.getByTestId('diff-metric-cell');
      expect(metricCell).toHaveClass('is-more');
    });

    it('applies is-less class when a > b', () => {
      render(<DiffNode {...baseProps} a={200} b={100} />);
      const metricCell = screen.getByTestId('diff-metric-cell');
      expect(metricCell).toHaveClass('is-less');
    });

    it('applies is-added class when a is 0', () => {
      render(<DiffNode {...baseProps} a={0} b={100} />);
      const metricCell = screen.getByTestId('diff-metric-cell');
      expect(metricCell).toHaveClass('is-added');
    });

    it('applies is-removed class when b is 0', () => {
      render(<DiffNode {...baseProps} a={100} b={0} />);
      const metricCell = screen.getByTestId('diff-metric-cell');
      expect(metricCell).toHaveClass('is-removed');
    });

    it('accepts unused isUiFindMatch prop without breaking', () => {
      render(<DiffNode {...baseProps} isUiFindMatch />);
      expect(screen.getByRole('table')).toBeInTheDocument();
    });
  });

  describe('renderNode()', () => {
    const lenA = 3;
    const lenB = 7;
    const key = 'vertex-key';
    const vertex = {
      data: {
        a: new Array(lenA),
        b: new Array(lenB),
        operation,
        service,
      },
      key,
    };

    it('extracts values from vertex.data and passes them as props', () => {
      const node = renderNode(vertex);
      expect(node.props.a).toBe(lenA);
      expect(node.props.b).toBe(lenB);
      expect(node.props.operation).toBe(operation);
      expect(node.props.service).toBe(service);
    });
  });

  describe('getNodeEmphasisRenderer', () => {
    const matchKey = 'match-key';
    const nonMatchKey = 'no-match';
    const renderer = getNodeEmphasisRenderer(new Set([matchKey]));

    it('returns EmphasizedNode when key matches', () => {
      const lv = { height: 100, width: 200, vertex: { key: matchKey } };
      const result = renderer(lv);
      expect(result.type).toBe(EmphasizedNode);
      expect(result.props.height).toBe(100);
      expect(result.props.width).toBe(200);
    });

    it('returns null when key does not match', () => {
      const lv = { height: 100, width: 200, vertex: { key: nonMatchKey } };
      const result = renderer(lv);
      expect(result).toBeNull();
    });
  });
});
