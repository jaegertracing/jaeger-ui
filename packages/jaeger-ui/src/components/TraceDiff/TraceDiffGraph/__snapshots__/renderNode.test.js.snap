// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`drawNode diffNode renders as expected when props.a and props.b are the same 1`] = `
<Popover
  content={
    <table
      className="DiffNode is-same"
    >
      <tbody
        className="DiffNode--body"
      >
        <tr>
          <td
            className="DiffNode--metricCell is-same"
            rowSpan={2}
          >
            100
          </td>
          <td
            className="DiffNode--labelCell is-same"
          >
            <strong>
              serviceName
            </strong>
            <CopyIcon
              className="DiffNode--copyIcon"
              copyText="serviceName operationName"
              icon="copy"
              placement="top"
              tooltipTitle="Copy label"
            />
          </td>
        </tr>
        <tr>
          <td
            className="DiffNode--labelCell is-same"
          >
            operationName
          </td>
        </tr>
      </tbody>
    </table>
  }
  mouseEnterDelay={0.25}
  overlayClassName="DiffNode--popover is-same"
>
  <table
    className="DiffNode is-same"
  >
    <tbody
      className="DiffNode--body"
    >
      <tr>
        <td
          className="DiffNode--metricCell is-same"
          rowSpan={2}
        >
          100
        </td>
        <td
          className="DiffNode--labelCell is-same"
        >
          <strong>
            serviceName
          </strong>
          <CopyIcon
            className="DiffNode--copyIcon"
            copyText="serviceName operationName"
            icon="copy"
            placement="top"
            tooltipTitle="Copy label"
          />
        </td>
      </tr>
      <tr>
        <td
          className="DiffNode--labelCell is-same"
        >
          operationName
        </td>
      </tr>
    </tbody>
  </table>
</Popover>
`;

exports[`drawNode diffNode renders as expected when props.a is 0 1`] = `
<Popover
  content={
    <table
      className="DiffNode is-changed is-added"
    >
      <tbody
        className="DiffNode--body"
      >
        <tr>
          <td
            className="DiffNode--metricCell is-changed is-added"
            rowSpan={1}
          >
            <span
              className="DiffNode--metricSymbol"
            >
              +
            </span>
            100
          </td>
          <td
            className="DiffNode--labelCell is-changed is-added"
          >
            <strong>
              serviceName
            </strong>
            <CopyIcon
              className="DiffNode--copyIcon"
              copyText="serviceName operationName"
              icon="copy"
              placement="top"
              tooltipTitle="Copy label"
            />
          </td>
        </tr>
        <tr>
          <td
            className="DiffNode--metricCell is-changed is-added"
          >
            <span
              className="DiffNode--metricSymbol"
            >
              +
            </span>
            100
            <span
              className="DiffNode--metricSymbol"
            >
              %
            </span>
          </td>
          <td
            className="DiffNode--labelCell is-changed is-added"
          >
            operationName
          </td>
        </tr>
      </tbody>
    </table>
  }
  mouseEnterDelay={0.25}
  overlayClassName="DiffNode--popover is-changed is-added"
>
  <table
    className="DiffNode is-changed is-added"
  >
    <tbody
      className="DiffNode--body"
    >
      <tr>
        <td
          className="DiffNode--metricCell is-changed is-added"
          rowSpan={1}
        >
          <span
            className="DiffNode--metricSymbol"
          >
            +
          </span>
          100
        </td>
        <td
          className="DiffNode--labelCell is-changed is-added"
        >
          <strong>
            serviceName
          </strong>
          <CopyIcon
            className="DiffNode--copyIcon"
            copyText="serviceName operationName"
            icon="copy"
            placement="top"
            tooltipTitle="Copy label"
          />
        </td>
      </tr>
      <tr>
        <td
          className="DiffNode--metricCell is-changed is-added"
        >
          <span
            className="DiffNode--metricSymbol"
          >
            +
          </span>
          100
          <span
            className="DiffNode--metricSymbol"
          >
            %
          </span>
        </td>
        <td
          className="DiffNode--labelCell is-changed is-added"
        >
          operationName
        </td>
      </tr>
    </tbody>
  </table>
</Popover>
`;

exports[`drawNode diffNode renders as expected when props.a is less than props.b 1`] = `
<Popover
  content={
    <table
      className="DiffNode is-changed is-more"
    >
      <tbody
        className="DiffNode--body"
      >
        <tr>
          <td
            className="DiffNode--metricCell is-changed is-more"
            rowSpan={1}
          >
            <span
              className="DiffNode--metricSymbol"
            >
              +
            </span>
            50
          </td>
          <td
            className="DiffNode--labelCell is-changed is-more"
          >
            <strong>
              serviceName
            </strong>
            <CopyIcon
              className="DiffNode--copyIcon"
              copyText="serviceName operationName"
              icon="copy"
              placement="top"
              tooltipTitle="Copy label"
            />
          </td>
        </tr>
        <tr>
          <td
            className="DiffNode--metricCell is-changed is-more"
          >
            <span
              className="DiffNode--metricSymbol"
            >
              +
            </span>
            50
            <span
              className="DiffNode--metricSymbol"
            >
              %
            </span>
          </td>
          <td
            className="DiffNode--labelCell is-changed is-more"
          >
            operationName
          </td>
        </tr>
      </tbody>
    </table>
  }
  mouseEnterDelay={0.25}
  overlayClassName="DiffNode--popover is-changed is-more"
>
  <table
    className="DiffNode is-changed is-more"
  >
    <tbody
      className="DiffNode--body"
    >
      <tr>
        <td
          className="DiffNode--metricCell is-changed is-more"
          rowSpan={1}
        >
          <span
            className="DiffNode--metricSymbol"
          >
            +
          </span>
          50
        </td>
        <td
          className="DiffNode--labelCell is-changed is-more"
        >
          <strong>
            serviceName
          </strong>
          <CopyIcon
            className="DiffNode--copyIcon"
            copyText="serviceName operationName"
            icon="copy"
            placement="top"
            tooltipTitle="Copy label"
          />
        </td>
      </tr>
      <tr>
        <td
          className="DiffNode--metricCell is-changed is-more"
        >
          <span
            className="DiffNode--metricSymbol"
          >
            +
          </span>
          50
          <span
            className="DiffNode--metricSymbol"
          >
            %
          </span>
        </td>
        <td
          className="DiffNode--labelCell is-changed is-more"
        >
          operationName
        </td>
      </tr>
    </tbody>
  </table>
</Popover>
`;

exports[`drawNode diffNode renders as expected when props.a is more than props.b 1`] = `
<Popover
  content={
    <table
      className="DiffNode is-changed is-less"
    >
      <tbody
        className="DiffNode--body"
      >
        <tr>
          <td
            className="DiffNode--metricCell is-changed is-less"
            rowSpan={1}
          >
            <span
              className="DiffNode--metricSymbol"
            >
              -
            </span>
            100
          </td>
          <td
            className="DiffNode--labelCell is-changed is-less"
          >
            <strong>
              serviceName
            </strong>
            <CopyIcon
              className="DiffNode--copyIcon"
              copyText="serviceName operationName"
              icon="copy"
              placement="top"
              tooltipTitle="Copy label"
            />
          </td>
        </tr>
        <tr>
          <td
            className="DiffNode--metricCell is-changed is-less"
          >
            <span
              className="DiffNode--metricSymbol"
            >
              -
            </span>
            50
            <span
              className="DiffNode--metricSymbol"
            >
              %
            </span>
          </td>
          <td
            className="DiffNode--labelCell is-changed is-less"
          >
            operationName
          </td>
        </tr>
      </tbody>
    </table>
  }
  mouseEnterDelay={0.25}
  overlayClassName="DiffNode--popover is-changed is-less"
>
  <table
    className="DiffNode is-changed is-less"
  >
    <tbody
      className="DiffNode--body"
    >
      <tr>
        <td
          className="DiffNode--metricCell is-changed is-less"
          rowSpan={1}
        >
          <span
            className="DiffNode--metricSymbol"
          >
            -
          </span>
          100
        </td>
        <td
          className="DiffNode--labelCell is-changed is-less"
        >
          <strong>
            serviceName
          </strong>
          <CopyIcon
            className="DiffNode--copyIcon"
            copyText="serviceName operationName"
            icon="copy"
            placement="top"
            tooltipTitle="Copy label"
          />
        </td>
      </tr>
      <tr>
        <td
          className="DiffNode--metricCell is-changed is-less"
        >
          <span
            className="DiffNode--metricSymbol"
          >
            -
          </span>
          50
          <span
            className="DiffNode--metricSymbol"
          >
            %
          </span>
        </td>
        <td
          className="DiffNode--labelCell is-changed is-less"
        >
          operationName
        </td>
      </tr>
    </tbody>
  </table>
</Popover>
`;

exports[`drawNode diffNode renders as expected when props.b is 0 1`] = `
<Popover
  content={
    <table
      className="DiffNode is-changed is-removed"
    >
      <tbody
        className="DiffNode--body"
      >
        <tr>
          <td
            className="DiffNode--metricCell is-changed is-removed"
            rowSpan={1}
          >
            <span
              className="DiffNode--metricSymbol"
            >
              -
            </span>
            100
          </td>
          <td
            className="DiffNode--labelCell is-changed is-removed"
          >
            <strong>
              serviceName
            </strong>
            <CopyIcon
              className="DiffNode--copyIcon"
              copyText="serviceName operationName"
              icon="copy"
              placement="top"
              tooltipTitle="Copy label"
            />
          </td>
        </tr>
        <tr>
          <td
            className="DiffNode--metricCell is-changed is-removed"
          >
            <span
              className="DiffNode--metricSymbol"
            >
              -
            </span>
            100
            <span
              className="DiffNode--metricSymbol"
            >
              %
            </span>
          </td>
          <td
            className="DiffNode--labelCell is-changed is-removed"
          >
            operationName
          </td>
        </tr>
      </tbody>
    </table>
  }
  mouseEnterDelay={0.25}
  overlayClassName="DiffNode--popover is-changed is-removed"
>
  <table
    className="DiffNode is-changed is-removed"
  >
    <tbody
      className="DiffNode--body"
    >
      <tr>
        <td
          className="DiffNode--metricCell is-changed is-removed"
          rowSpan={1}
        >
          <span
            className="DiffNode--metricSymbol"
          >
            -
          </span>
          100
        </td>
        <td
          className="DiffNode--labelCell is-changed is-removed"
        >
          <strong>
            serviceName
          </strong>
          <CopyIcon
            className="DiffNode--copyIcon"
            copyText="serviceName operationName"
            icon="copy"
            placement="top"
            tooltipTitle="Copy label"
          />
        </td>
      </tr>
      <tr>
        <td
          className="DiffNode--metricCell is-changed is-removed"
        >
          <span
            className="DiffNode--metricSymbol"
          >
            -
          </span>
          100
          <span
            className="DiffNode--metricSymbol"
          >
            %
          </span>
        </td>
        <td
          className="DiffNode--labelCell is-changed is-removed"
        >
          operationName
        </td>
      </tr>
    </tbody>
  </table>
</Popover>
`;

exports[`drawNode diffNode renders as expected when props.isUiFindMatch is true 1`] = `
<Popover
  content={
    <table
      className="DiffNode is-same"
    >
      <tbody
        className="DiffNode--body"
      >
        <tr>
          <td
            className="DiffNode--metricCell is-same"
            rowSpan={2}
          >
            100
          </td>
          <td
            className="DiffNode--labelCell is-same"
          >
            <strong>
              serviceName
            </strong>
            <CopyIcon
              className="DiffNode--copyIcon"
              copyText="serviceName operationName"
              icon="copy"
              placement="top"
              tooltipTitle="Copy label"
            />
          </td>
        </tr>
        <tr>
          <td
            className="DiffNode--labelCell is-same"
          >
            operationName
          </td>
        </tr>
      </tbody>
    </table>
  }
  mouseEnterDelay={0.25}
  overlayClassName="DiffNode--popover is-same"
>
  <table
    className="DiffNode is-same"
  >
    <tbody
      className="DiffNode--body"
    >
      <tr>
        <td
          className="DiffNode--metricCell is-same"
          rowSpan={2}
        >
          100
        </td>
        <td
          className="DiffNode--labelCell is-same"
        >
          <strong>
            serviceName
          </strong>
          <CopyIcon
            className="DiffNode--copyIcon"
            copyText="serviceName operationName"
            icon="copy"
            placement="top"
            tooltipTitle="Copy label"
          />
        </td>
      </tr>
      <tr>
        <td
          className="DiffNode--labelCell is-same"
        >
          operationName
        </td>
      </tr>
    </tbody>
  </table>
</Popover>
`;
